src/ppx_variants_conv.ml

```diff
@@ -4,619 +4,830 @@
-    if Option.is_some cd.pcd_res then
-      Location.raise_errorf ~loc:cd.pcd_loc "GADTs are not supported by variantslib";
-    let kind =
-      match cd.pcd_args with
-      | Pcstr_tuple pcd_args -> `Normal pcd_args
-      | Pcstr_record fields  -> `Normal_inline_record fields
-    in
-    { name = cd.pcd_name.txt
-    ; loc = cd.pcd_name.loc
-    ; kind
-    }
-
-  let type_decl td =
-    let loc = td.ptype_loc in
-    match td.ptype_kind with
-    | Ptype_variant cds ->
-      let cds = List.map cds ~f:constructor in
-      let names_as_string = Hashtbl.create (module String) in
-      List.iter cds ~f:(fun { name; loc; _ } ->
-        let s = variant_name_to_string name in
-        match Hashtbl.find names_as_string s with
-        | None -> Hashtbl.add_exn names_as_string ~key:s ~data:name
-        | Some name' ->
+   open Base
+   open Ppxlib
+   open Ast_builder.Default
+   
+   let raise_unsupported loc =
+     Location.raise_errorf ~loc
+       "Unsupported use of variants (you can only use it on variant types)."
+
+   let raise_unsupported_ty loc =
+     Location.raise_errorf ~loc
+       "Use of this type is not supported use by variants."
+   
+   module Create = struct
+     let lambda loc xs body =
+       List.fold_right xs ~init:body ~f:(fun (label, p) e -> pexp_fun ~loc label None p e)
+     ;;
+   
+     let lambda_sig loc arg_tys body_ty =
+       List.fold_right arg_tys ~init:body_ty ~f:(fun (label, arg_ty) acc ->
+         ptyp_arrow ~loc label arg_ty acc)
+     ;;
+   end
+
+   module Core_type = struct
+
+     let tyvar_name i =
+       let c, n = (Int.rem i 26, i / 26) in
+       let a = String.of_char @@ Char.of_int_exn (c + 97)
+       and v = if n = 0 then "" else Int.to_string n in
+       a ^ v
+     ;;
+
+     let (supply, reset_supply) =
+       let n = ref 0 in
+       ((fun () ->
+         let nm = tyvar_name !n in
+         Int.incr n ; nm)
+       ,(fun () -> n := 0))
+     ;;
+
+     let supply_excluding bound = 
+       let rec aux () = 
+         let next = supply () in
+         if List.exists ~f:(String.equal next) bound then aux ()
+         else next
+       in
+       aux 
+    ;; 
+
+     (** Replace wildcard type parameter with fresh type variables *)
+     let canonical_tycon td = 
+       let loc = td.ptype_loc in
+       let params = 
+         let all = List.map ~f:fst td.ptype_params in
+         let bound = 
+           List.filter_map all ~f:(fun core_ty ->
+             match core_ty.ptyp_desc with
+             | Ptyp_var v -> Some v 
+             | Ptyp_any -> None 
+             | _ -> assert(false)) in
+         let _ : unit = reset_supply () in
+         let fresh_var = supply_excluding bound in
+         List.map all ~f:(fun core_ty ->
+           match core_ty.ptyp_desc with
+           | Ptyp_any -> {core_ty with ptyp_desc = Ptyp_var (fresh_var ()) }
+           | _ -> core_ty
+         )
+       in
+       ptyp_constr ~loc {loc;txt= Longident.Lident td.ptype_name.txt } params 
```

Comment:
Pretty sure this is ppxlib.name_type_params_in_td

Reply:
Thanks again - I wasn't aware of these helpers